# Pandas Dataframe

<img src="https://frenzy86.s3.eu-west-2.amazonaws.com/IFAO/dataframe.png" width="1200">

In [13]:
import numpy as np
import pandas as pd

In [14]:
np.random.seed(667)
mydata = np.random.randint(0,101,(4,3))

In [15]:
mydata

array([[ 85,  15,  89],
       [100,   3,  90],
       [  2,  64,   0],
       [ 36,  99,  47]])

In [16]:
myindex = ['CA','NY','AZ','TX']

In [17]:
mycolumns = ['Jan','Feb','Mar']

In [18]:
df = pd.DataFrame(data=mydata)
df

0   1   2
0   85  15  89
1  100   3  90
2    2  64   0
3   36  99  47

In [19]:
df = pd.DataFrame(data=mydata,index=myindex)
df

0   1   2
CA   85  15  89
NY  100   3  90
AZ    2  64   0
TX   36  99  47

In [20]:
df = pd.DataFrame(myindex,mycolumns,mydata)


ValueError: ignored

In [21]:
df = pd.DataFrame(data=mydata,index=myindex,columns=mycolumns)
df 

Jan  Feb  Mar
CA   85   15   89
NY  100    3   90
AZ    2   64    0
TX   36   99   47

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, CA to TX
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Jan     4 non-null      int64
 1   Feb     4 non-null      int64
 2   Mar     4 non-null      int64
dtypes: int64(3)
memory usage: 128.0+ bytes


## Read external data tables

In [23]:
!wget https://frenzy86.s3.eu-west-2.amazonaws.com/IFAO/tips_complete.csv

--2023-03-31 12:11:08--  https://frenzy86.s3.eu-west-2.amazonaws.com/IFAO/tips_complete.csv
Resolving frenzy86.s3.eu-west-2.amazonaws.com (frenzy86.s3.eu-west-2.amazonaws.com)... 52.95.142.118
Connecting to frenzy86.s3.eu-west-2.amazonaws.com (frenzy86.s3.eu-west-2.amazonaws.com)|52.95.142.118|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18752 (18K) [text/csv]
Saving to: ‘tips_complete.csv’

tips_complete.csv   100%[===================>]  18.31K  74.6KB/s    in 0.2s    

2023-03-31 12:11:09 (74.6 KB/s) - ‘tips_complete.csv’ saved [18752/18752]



In [24]:
df = pd.read_csv('tips_complete.csv')
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

----
About this DataSet (in case you are interested)

* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:

* Format
    * A data frame with 244 rows and 7 variables

* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.

In all he recorded 244 tips.
    
* Note: I created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

# DataFrames

In [26]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [27]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [28]:
df.head(10)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   
5       25.29  4.71    Male     No  Sun  Dinner     4              6.32   
6        8.77  2.00    Male     No  Sun  Dinner     2              4.38   
7       26.88  3.12    Male     No  Sun  Dinner     4              6.72   
8       15.04  1.96    Male     No  Sun  Dinner     2              7.52   
9       14.78  3.23    Male     No  Sun  Dinner     2              7.39   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251  
5          Erik Smith   213140353657882    Sun9679  
6  Kristopher Johnson  2223727524230344    Sun5985  
7         Robert Buck  3514785077705092    Sun8157  
8     Joseph Mcdonald  3522866365840377    Sun6820  
9       Jerome Abbott  3532124519049786    Sun3775

In [29]:
df.tail(3)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


In [31]:
len(df)

244

In [33]:
df.describe().T

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

In [ ]:
#df.describe().transpose()
df.describe().T

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

## Selection and Indexing

Let's learn how to retrieve information from a DataFrame.

### COLUMNS

We will begin be learning how to extract information based on the columns

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

#### Grab a Single Column

In [34]:
df['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [35]:
type(df['total_bill'])

pandas.core.series.Series

In [36]:
df.total_bill

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

#### Grab Multiple Columns

In [38]:
# Note how its a python list of column names! Thus the double brackets.
df[['total_bill','tip']] #è una lista di selezione!!!

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

#### Create New Columns

In [39]:
df['tip_percentage'] = 100* df['tip'] / df['total_bill']

In [40]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959        5.944673  
1        Douglas Tucker  4478071379779230    Sun4608       16.054159  
2        Travis Walters  6011812112971322    Sun4458       16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4          Tonya Carter  4832732618637221    Sun2251       14.680765  
..                  ...               ...        ...             ...  
239       Michael Avila  5296068606052842    Sat2657       20.392697  
240      Monica Sanders  3506806155565404    Sat1766        7.358352  
241          Keith Wong  6011891618747196    Sat3880        8.822232  
242        Dennis Dixon     4375220550950      Sat17        9.820426  
243     Michelle Hardin  3511451626698139    Thur672       15.974441  

[244 rows x 12 columns]

In [ ]:
df['price_per_person'] = df['total_bill'] / df['size']

In [ ]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2          8.495000   
1         10.34  1.66    Male     No   Sun  Dinner     3          3.446667   
2         21.01  3.50    Male     No   Sun  Dinner     3          7.003333   
3         23.68  3.31    Male     No   Sun  Dinner     2         11.840000   
4         24.59  3.61  Female     No   Sun  Dinner     4          6.147500   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3          9.676667   
240       27.18  2.00  Female    Yes   Sat  Dinner     2         13.590000   
241       22.67  2.00    Male    Yes   Sat  Dinner     2         11.335000   
242       17.82  1.75    Male     No   Sat  Dinner     2          8.910000   
243       18.78  3.00  Female     No  Thur  Dinner     2          9.390000   

             Payer Name         CC Number Payment ID  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959        5.944673  
1        Douglas Tucker  4478071379779230    Sun4608       16.054159  
2        Travis Walters  6011812112971322    Sun4458       16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4          Tonya Carter  4832732618637221    Sun2251       14.680765  
..                  ...               ...        ...             ...  
239       Michael Avila  5296068606052842    Sat2657       20.392697  
240      Monica Sanders  3506806155565404    Sat1766        7.358352  
241          Keith Wong  6011891618747196    Sat3880        8.822232  
242        Dennis Dixon     4375220550950      Sat17        9.820426  
243     Michelle Hardin  3511451626698139    Thur672       15.974441  

[244 rows x 12 columns]

In [41]:
help(np.round)

Help on function round_ in module numpy:

round_(a, decimals=0, out=None)
    Round an array to the given number of decimals.
    
    See Also
    --------
    around : equivalent function; see for details.



#### Adjust Existing Columns

In [42]:
# Because pandas is based on numpy, we get awesome capabilities with numpy's universal functions!
df['price_per_person'] = np.round(df['price_per_person'],2)

In [43]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

#### Remove Columns

In [44]:
df.drop('tip_percentage',axis=1)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

In [45]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959        5.944673  
1        Douglas Tucker  4478071379779230    Sun4608       16.054159  
2        Travis Walters  6011812112971322    Sun4458       16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4          Tonya Carter  4832732618637221    Sun2251       14.680765  
..                  ...               ...        ...             ...  
239       Michael Avila  5296068606052842    Sat2657       20.392697  
240      Monica Sanders  3506806155565404    Sat1766        7.358352  
241          Keith Wong  6011891618747196    Sat3880        8.822232  
242        Dennis Dixon     4375220550950      Sat17        9.820426  
243     Michelle Hardin  3511451626698139    Thur672       15.974441  

[244 rows x 12 columns]

In [46]:
df = df.drop("tip_percentage",axis=1)

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

# Index Basics

Before going over the same retrieval tasks for rows, let's build some basic understanding of the pandas DataFrame Index.

In [47]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [48]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [49]:
df.set_index('Payment ID')

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221  
...                      ...                 ...               ...  
Sat2657                 9.68       Michael Avila  5296068606052842  
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Thur672                 9.39     Michelle Hardin  3511451626698139  

[244 rows x 10 columns]

In [50]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [51]:
df = df.set_index('Payment ID')

In [52]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [53]:
df = df.reset_index()

In [54]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0              8.49  Christy Cunningham  3560325168603410  
1              3.45      Douglas Tucker  4478071379779230  
2              7.00      Travis Walters  6011812112971322  
3             11.84    Nathaniel Harris  4676137647685994  
4              6.15        Tonya Carter  4832732618637221

### ROWS

Let's now explore these same concepts but with Rows.

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [55]:
df = df.set_index('Payment ID')
df

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221  
...                      ...                 ...               ...  
Sat2657                 9.68       Michael Avila  5296068606052842  
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Thur672                 9.39     Michelle Hardin  3511451626698139  

[244 rows x 10 columns]

#### Grab a Single Row

In [56]:
# Integer Based
df.iloc[0]

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [58]:
# Name Based
df.loc['Sun4608']

total_bill                     10.34
tip                             1.66
sex                             Male
smoker                            No
day                              Sun
time                          Dinner
size                               3
price_per_person                3.45
Payer Name            Douglas Tucker
CC Number           4478071379779230
Name: Sun4608, dtype: object

#### Grab Multiple Rows

In [60]:
df.iloc[0:4] #prime 4 righe

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994

In [67]:
df.loc[['Sun2959','Sun5260']]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun5260                11.84    Nathaniel Harris  4676137647685994

#### Remove Row

Typically are datasets will be large enough that we won't remove rows like this since we won't know thier row location for some specific condition, instead, we drop rows based on conditions such as missing data or column values. The next lecture will cover this in a lot more detail.

In [68]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [69]:
df.drop('Sun2959',axis=0).head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number  
Payment ID                                                        
Sun4608                 3.45    Douglas Tucker  4478071379779230  
Sun4458                 7.00    Travis Walters  6011812112971322  
Sun5260                11.84  Nathaniel Harris  4676137647685994  
Sun2251                 6.15      Tonya Carter  4832732618637221  
Sun9679                 6.32        Erik Smith   213140353657882

In [ ]:
# Error if you have a named index!
# df.drop(0,axis=0).head()

#### Insert a New Row

Pretty rare to add a single row like this. Usually you use pd.concat() to add many rows at once. You could use the .append() method with a list of pd.Series() objects, but you won't see us do this with realistic real-world data.

In [70]:
one_row = df.iloc[0]

In [71]:
one_row

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [ ]:
type(one_row)

pandas.core.series.Series

In [ ]:
df.tail()

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person       Payer Name         CC Number  
Payment ID                                                       
Sat2657                 9.68    Michael Avila  5296068606052842  
Sat1766                13.59   Monica Sanders  3506806155565404  
Sat3880                11.34       Keith Wong  6011891618747196  
Sat17                   8.91     Dennis Dixon     4375220550950  
Thur672                 9.39  Michelle Hardin  3511451626698139

In [72]:
df.append(one_row).tail()

<ipython-input-72-161d04e87f2b>:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df.append(one_row).tail()


total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Thur672                 9.39     Michelle Hardin  3511451626698139  
Sun2959                 8.49  Christy Cunningham  3560325168603410

--------

#### Extract an array

In [73]:
array_row = df.iloc[0].values
array_row

array([16.99, 1.01, 'Female', 'No', 'Sun', 'Dinner', 2, 8.49,
       'Christy Cunningham', 3560325168603410], dtype=object)

In [74]:
df['total_'].tolist()

KeyError: ignored